crates/sui-types/src/storage/error.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+pub type Result<T, E = Error> = ::std::result::Result<T, E>;
+
+#[derive(Debug)]
+pub struct Error {
+    inner: Box<Inner>,
+}
+
+type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+#[derive(Debug)]
+struct Inner {
+    kind: Kind,
+    source: Option<BoxError>,
```

Comment:
string

Reply:
I don't think having an explicit "origin" is very beneficial for a few reasons:
1. It requires that we ahead of time know these "origins" which sort of couples this generic interface with any backends.
2. Its not really needed since the origin error is curried along in the source field, so any concrete error information can be included there.
3. If we really want some true form of "origin" we could add a standard backtrace field to the error type to capture a backtrace to where the error occurred which would be more accurate than some "origin" field imo.